### PR TITLE
[ESQL] Move binary operator compatbility checking into type resolution

### PR DESCRIPTION
**Pull request by not-napoleon**
This starts moving the binary compatibility type checking out of the verifier and into the type resolve path.  Additionally, and as a test case, it wires up the addition tests to use `errorForCasesWithoutExamples`, which requires the type resolve improvement, and validates that we're doing the right thing for binary addition.

Before we can actually remove the code from the verifier, we need to also adapt the binary comparison functions, which is a fair bit of work.  I am planning to do that in a follow up PR, as this one has already been open for two weeks.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/EsqlArithmeticOperation.java [hunk 1]
```diff
@@ -109,6 +112,49 @@ public DataType dataType() {
...
+    protected TypeResolution checkCompatibility() {
+        // This checks that unsigned longs should only be compatible with other unsigned longs
+        DataType leftType = left().dataType();
+        DataType rightType = right().dataType();
+        if ((rightType == DataTypes.UNSIGNED_LONG && (false == (leftType == UNSIGNED_LONG || leftType == DataTypes.NULL)))
+            || (leftType == DataTypes.UNSIGNED_LONG && (false == (rightType == UNSIGNED_LONG || rightType == DataTypes.NULL)))) {
+            return new TypeResolution(formatIncombatibleTypesMessage(symbol(), leftType.typeName(), rightType.typeName()));
+        }
+        // If the LHS is numeric, the RHS should be numeric or null
+        if (leftType.isNumeric()) {
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/EsqlArithmeticOperation.java:139]: I'm wondering if we need this check at all...
Before we reach this point, `resolveType()` already checked that both arguments are numbers or dates, and for `DateTimeArithmeticOperation` we have an override that checks that they are compatible.
Maybe I'm missing some execution paths, but in that case we should probably have a symmetric check here, ie. `else if(rightType.isNumeric()) { //check left type`

**Review by luigidellaquila — APPROVED** [review]
Thanks Mark, LGTM.
I just left a comment about a check that is probably redundant (or maybe I'm missing something)

**Review by astefan — COMMENTED** [review]
Left some comments (mostly cosmetical), one of which is about error messages consistency.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/EsqlArithmeticOperation.java (hunk 1 above)
Comment (not-napoleon) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/EsqlArithmeticOperation.java:139]: Yeah, I think you're right, this is redundant.  I'll remove it.  Good catch, thanks.

**Review by astefan — APPROVED** [review]
LGTM. Thank you, Mark

**Pull request merged**